File: Algorithms/copyListWithRandomPointer.py
class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        self.Clone(head)
        self.CopyRandomPointer(head)
        return self.restore(head)

    def Clone(self, head):
        pNode = head
        while pNode:
            clonedNode = RandomListNode(pNode.label)
            clonedNode.next = pNode.next
            pNode.next = clonedNode
            pNode = clonedNode.next

    def CopyRandomPointer(self, head):
        pNode = head
        while pNode:
            clonedNode = pNode.next
            if pNode.random:
                clonedNode.random = pNode.random.next
            pNode = clonedNode.next

    def restore(self, head):
        pNode = head
        pClonedHead = None
        pClonedNode = None
        if pNode:
            pClonedHead = pNode.next
            pClonedNode = pNode.next
            pNode.next = pClonedNode.next
            pNode = pNode.next
        while pNode:
            pClonedNode.next = pNode.next
            pClonedNode = pClonedNode.next
            pNode.next = pClonedNode.next
            pNode = pNode.next
        return pClonedHead

File: Algorithms/copyListWithRandomPointer.java
import java.util.*;
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        Clone(head);
        CopyRandomPointer(head);
        return restore(head);
    }
    void Clone(RandomListNode head) {
        RandomListNode pNode = head;
        while (pNode != null) {
            RandomListNode clonedNode = new RandomListNode(pNode.label);
            clonedNode.next = pNode.next;
            pNode.next = clonedNode;
            pNode = clonedNode.next;
        }
    }
    void CopyRandomPointer(RandomListNode head) {
        RandomListNode pNode = head;
        while (pNode != null) {
            RandomListNode clonedNode = pNode.next;
            if (pNode.random != null) {
                clonedNode.random = pNode.random.next;
            }
            pNode = clonedNode.next;
        }
    }
    RandomListNode restore(RandomListNode head) {
        RandomListNode pNode = head;
        RandomListNode pClonedHead = null;
        RandomListNode pClonedNode = null;
        if (pNode != null) {
            pClonedHead = pNode.next;
            pClonedNode = pNode.next;
            pNode.next = pClonedNode.next;
            pNode = pNode.next; 
        }
        while (pNode != null) {
            pClonedNode.next = pNode.next;
            pClonedNode = pClonedNode.next;
            pNode.next = pClonedNode.next;
            pNode = pNode.next;
        }
        return pClonedHead;
    }
}
